provider/sidecar_test.go

```diff
@@ -15,215 +14,241 @@ import (
+
 		So(err, ShouldBeNil)
 		So(configMsg.ProviderName, ShouldEqual, "sidecar")
 		So(configMsg.Configuration.Frontends["web"].Routes["test_1"].Rule, ShouldEqual, "Host: some-aws-host")
-		So(configMsg.Configuration.Backends["web"].Servers["some-aws-host"].URL, ShouldEndWith, "http://some-aws-host:8000")
+		So(configMsg.Configuration.Backends["web"].Servers["some-aws-host"].URL, ShouldEqual, "http://some-aws-host:8000")
 	})
 }
 
-func Test_SidecarWatcher(t *testing.T) {
-	Convey("Verify Sidecar Provider", t, func() {
-		httpmock.Activate()
-		defer httpmock.DeactivateAndReset()
+func TestSidecarWatcher(t *testing.T) {
+	setup()
+	defer teardown()
 
-		httpmock.RegisterResponder("GET", "http://some.dummy.service/state.json",
+	Convey("Verify Sidecar Watcher", t, func(c C) {
+		// Return the watcher mock HTTP response on demand
+		releaseWatch := make(chan bool)
+		httpmock.RegisterResponder(http.MethodGet, "http://some.dummy.service/watch",
 			func(req *http.Request) (*http.Response, error) {
+				<-releaseWatch
 
-				testPort := service.Port{Type: "tcp", Port: 9000, ServicePort: 9000}
-				returnState := catalog.NewServicesState()
-				baseTime := time.Now().UTC().Round(time.Second)
-				serv := service.Service{ID: "007", Name: "web", Hostname: "some-aws-host",
-					Updated: baseTime.Add(5 * time.Second), Status: 0, Ports: []service.Port{testPort}}
-				returnState.AddServiceEntry(serv)
-				resp, err := httpmock.NewJsonResponse(200, returnState)
+				resp, err := httpmock.NewJsonResponse(200, dummyState.ByService())
 				if err != nil {
 					return httpmock.NewStringResponse(500, ""), nil
 				}
 				return resp, nil
 			},
 		)
 
-		httpmock.RegisterResponder("GET", "http://some.dummy.service/watch",
-			func(req *http.Request) (*http.Response, error) {
-
-				testPort := service.Port{Type: "tcp", Port: 9000, ServicePort: 9000}
-				returnState := catalog.NewServicesState()
-				baseTime := time.Now().UTC().Round(time.Second)
-				serv := service.Service{ID: "007", Name: "web", Hostname: "some-aws-host",
-					Updated: baseTime.Add(5 * time.Second), Status: 0, Ports: []service.Port{testPort}}
-				returnState.AddServiceEntry(serv)
-				resp, err := httpmock.NewJsonResponse(200, returnState.ByService())
-				if err != nil {
-					return httpmock.NewStringResponse(500, ""), nil
-				}
-				return resp, nil
-			},
-		)
 		prov := Sidecar{
+			BaseProvider: BaseProvider{
+				Watch: true,
+			},
 			Endpoint: "http://some.dummy.service",
+			Frontend: "testdata/sidecar_testdata.toml",
 		}
-		prov.Watch = true
-		prov.Frontend = "testdata/sidecar_testdata.toml"
-		configurationChan := make(chan types.ConfigMessage, 100)
-		constraints := types.Constraints{}
-		pool := safe.NewPool(context.Background())
-		go prov.Provide(configurationChan, pool, constraints)
-		configMsg, _ := <-configurationChan
+
+		// We have no way to shut down the provider when it's running in watch mode,
+		// so just let the unit test close it at the end
```

Comment:
@relistan I have pushed 4635181 to address this comment. Please let me know if that change fully addresses this code smell.